### ISSUE TITLE
Numerical inaccuracies in "ddp_apply_optim_in_backward" unit tests

### ISSUE THREAD

**Issue by jataylo** [pytorch/pytorch#111834]
### 🐛 Describe the bug

After some experiments in https://github.com/pytorch/pytorch/pull/111791 I have replicated an accuracy issue with the gloo backend relating to ddp models using "apply_optim_in_backwards" instead of .step() on the CI. This occurs both for CUDA and ROCm.

There are already unit tests in https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/distributed/distributed_test.py that track this behavior but I have found that torchvision is not present causing these tests to only run for a simple linear model in which the bug is not present.

This can be replicated with TOT PyTorch with the following unit test:
`BACKEND=gloo WORLD_SIZE=2 PYTORCH_TEST_WITH_ROCM=1 HIP_VISIBLE_DEVICES=0,1 python3 test/distributed/test_distributed_spawn.py TestDistBackendWithSpawn.test_ddp_apply_optim_in_backward_grad_as_bucket_view_false`

### Versions

This can be replicated with both CUDA and ROCm CI environments in the distributed workflow if we modify the job to install torchvision as seen here:
https://hud.pytorch.org/pr/111791

**Comment by jataylo** [comment]
This seems to occur with even simple models as long as there is a convolution involved.

I updated the `test_ddp_apply_optim_in_backward` UT to test a dummy model and we can still replicate the issue
 ```
def _test_ddp_apply_optim_in_backward(
    self,
    optim_cls,
    optim_kwargs,
    init_before,
    gradient_as_bucket_view=True,
):
    # Need to seed to ensure inputs are unique across rank. Otherwise,
    # allreduce won't have any effect.
    torch.manual_seed(self.rank)
    torch.cuda.manual_seed(self.rank)
    torch.cuda.set_device(self.rank)

    # Define a toymodel
    class BasicModel(nn.Module):
        def __init__(self, use_bn=False, input_dim=512, input_channels=3, conv_out_channels=8, kernel_size=3, stride=1, padding=0, linear_out=32):
            super(BasicModel, self).__init__()
            self.use_bn = use_bn

            if self.use_bn:
                self.bn = nn.BatchNorm2d(input_channels)
                output_size = input_dim
                self.fc_input_size = input_channels * output_size * output_size
            else:
                self.conv = nn.Conv2d(input_channels, conv_out_channels, kernel_size, stride, padding)
                output_size = ((input_dim - kernel_size + 2 * padding) // stride) + 1  # Both width and height
                self.fc_input_size = conv_out_channels * output_size * output_size

            self.fc = nn.Linear(self.fc_input_size, linear_out)

        def forward(self, x):
            if self.use_bn:
                x = self.bn(x)   # Issue does not occur if we just perform a batch norm
            else:
                x = self.conv(x) # Issue occurs with a convolution involved
            x = x.view(x.size(0), -1)  # Flatten the tensor
            x = self.fc(x)
            return x

    # Your original code starts here
    models_to_test = []
    input_size=40
    models_to_test.append(BasicModel(input_dim=input_size, use_bn=False).cuda())

    for j, model in enumerate(models_to_test):
        model_optim_in_bwd = copy.deepcopy(model)
        model = nn.parallel.DistributedDataParallel(
            model,
            device_ids=[self.rank],
            gradient_as_bucket_view=gradient_as_bucket_view,
        )

        optim = optim_cls(model.parameters(), **optim_kwargs)

        _apply_optimizer_in_backward(
            optimizer_class=optim_cls,
            params=model_optim_in_bwd.parameters(),
            optimizer_kwargs=optim_kwargs,
            register_hook=False,
        )

        model_optim_in_bwd = nn.parallel.DistributedDataParallel(
            model_optim_in_bwd,
            device_ids=[self.rank],
            gradient_as_bucket_view=gradient_as_bucket_view,
        )

        for p1, p2 in zip(model.parameters(), model_optim_in_bwd.parameters()):
            self.assertEqual(p1, p2, "Parameters not initially equal!")

        # Enable determinism in cudnn operators
        with torch.backends.cudnn.flags(
            enabled=True, deterministic=True, benchmark=False
        ):
            for i in range(8):
                inp = (
                    torch.randn(1, 3, input_size, input_size, device="cuda")
                )

                model(inp).sum().backward()
                optim.step()
                model_optim_in_bwd(
                    inp
                ).sum().backward()  # runs optimizer as well

                print(f"Iteration: {i}")
                for p1, p2 in zip(
                    model.parameters(), model_optim_in_bwd.parameters()
                ):
                    self.assertEqual(
                        p1, p2, f"Params not equal at iteration {i}"
                    )

                    self.assertTrue(
                        p2.grad is None,
                        f"Optim in backward grad is not None at {i}",
                    )

                # set_to_none for regular optimizer to match in backward
                # case.
                optim.zero_grad(set_to_none=True)

```

**Comment by jithunnair-amd** [comment]
@wconstab This is a longstanding issue that we would like to get closure on.

**Comment by fduwjj** [comment]
@jithunnair-amd We had this fix recently https://github.com/pytorch/pytorch/pull/153862, does it help? also cc: @jataylo
